tools/functional-tester/etcd-tester/lease_stresser.go

```diff
@@ -0,0 +1,373 @@
+}
+
+func (ls *leaseStresser) Stress() error {
+	plog.Printf("lease Stresser %v starting ...", ls)
+	ctx, cancel := context.WithCancel(context.Background())
+	ls.cancel = cancel
+	ls.run(ctx)
+	return nil
+}
+
+func (ls *leaseStresser) run(ctx context.Context) {
+
+	ls.createLeases(ctx)
+	ls.randomlyDropLeases(ctx)
+
+}
+
+func (ls *leaseStresser) createLeases(ctx context.Context) {
+	numOfLeasesToCreate := ls.numOfLeases - len(ls.aliveLeases)
+	var wg sync.WaitGroup
+	wg.Add(numOfLeasesToCreate)
+	for i := 0; i < numOfLeasesToCreate; i++ {
+		go func() {
+			defer wg.Done()
+			leaseID, err := ls.createLeaseAndKeepAlive(ctx)
+
+			if err != nil {
+				return
+			}
+			plog.Printf("Lease %v created", leaseID)
+			ls.attachKeysWithLease(ctx, leaseID)
+			ls.mu.Lock()
+			ls.aliveLeases[leaseID] = struct{}{}
+			plog.Printf("Alive leases %v", ls.aliveLeases)
+			ls.mu.Unlock()
+
+		}()
+	}
+	wg.Wait()
+}
+
+func (ls *leaseStresser) randomlyDropLeases(ctx context.Context) {
+	var wg sync.WaitGroup
+	wg.Add(len(ls.aliveLeases))
+	for l := range ls.aliveLeases {
+		go func(leaseID int64) {
+			defer wg.Done()
+			isLeaseDropped := ls.randomlyDropLease(ctx, leaseID)
+			if isLeaseDropped {
+				ls.mu.Lock()
+				plog.Printf("Lease %v dropped", leaseID)
+				ls.revokedLeases[leaseID] = struct{}{}
+				delete(ls.aliveLeases, leaseID)
+				ls.mu.Unlock()
+			}
+		}(l)
+	}
+	wg.Wait()
+
+}
+
+func (ls *leaseStresser) InvariantCheck(ctx context.Context) error {
+	plog.Printf("Revoked Leases %+v ", ls.revokedLeases)
+	for leaseID := range ls.revokedLeases {
+		plog.Printf("Revoked Lease %v ", leaseID)
+		keysExpired, err := ls.hasKeysExpired(ctx, leaseID)
+		if err != nil {
+			return err
+		}
+
+		leaseExpired, err := ls.hasLeaseExpired(ctx, leaseID)
+		if err != nil {
+			return err
+		}
+
+		if !keysExpired || !leaseExpired {
+			return fmt.Errorf("revoked lease %v not expired", leaseID)
+		}
+	}
+
+	ls.mu.Lock()
+	ls.revokedLeases = make(map[int64]struct{})
+	ls.mu.Unlock()
+
+	// for those leases that are alive, checks whether if they still are after waiting for 2 * TTL
+	if len(ls.aliveLeases) != 0 {
+		time.Sleep(TTL * 2 * time.Second)
+	}
+
+	plog.Printf("Alive Leases %+v ", ls.aliveLeases)
+	for leaseID := range ls.aliveLeases {
+		keysExpired, err := ls.hasKeysExpired(ctx, leaseID)
+		if err != nil {
+			return err
+		}
+
+		leaseExpired, err := ls.hasLeaseExpired(ctx, leaseID)
+		if err != nil {
+			return err
+		}
+
+		if keysExpired || leaseExpired {
+			return fmt.Errorf("live lease %v expired", leaseID)
+		}
+	}
+	return nil
+
+}
+
+func (ls *leaseStresser) getLeaseByID(ctx context.Context, leaseID int64) (*pb.LeaseTimeToLiveResponse, error) {
+	LTL := &pb.LeaseTimeToLiveRequest{
+		ID:   leaseID,
+		Keys: true,
+	}
+	return ls.lc.LeaseTimeToLive(ctx, LTL, grpc.FailFast(false))
+}
+
+func (ls *leaseStresser) hasLeaseExpired(ctx context.Context, leaseID int64) (bool, error) {
+
+	_, err := ls.getLeaseByID(ctx, leaseID)
+
+	if err != nil {
```

Comment:
basically

server -> grpcError -> grpc -> client -> convert(grpcError) to error. Then you can compare error directly against what we defined.